core/domain/email_manager_test.py

```diff
@@ -426,3 +428,422 @@ def test_record_of_sent_email_is_written_to_datastore(self):
                 sent_email_model.subject, 'Welcome!')
             self.assertEqual(
                 sent_email_model.html_body, self.expected_html_email_content)
+
+
+class GenerateHashTests(test_utils.GenericTestBase):
+    """Test that generating hash functionality works as expected."""
+
+    def test_same_input_always_gives_same_hash(self):
+        # pylint: disable=protected-access
+        email_hash1 = email_manager._generate_hash('recipient_id',
+                                                   'email_subject',
+                                                   'email_html_body1')
+
+        email_hash2 = email_manager._generate_hash('recipient_id',
+                                                   'email_subject',
+                                                   'email_html_body1')
+        self.assertEqual(email_hash1, email_hash2)
+        # pylint: enable=protected-access
+
+    def test_different_input_give_different_hash(self):
+        # pylint: disable=protected-access
+        email_hash1 = email_manager._generate_hash('recipient_id',
+                                                   'email_subject',
+                                                   'email_html_body1')
```

Comment:
Nit: perhaps this should be email_html_body (it seems to make more sense, based on your checks below)